File: src/main/java/com/falsepattern/zigbrains/zon/completion/ZonCompletionContributor.java
/*
 * Copyright 2023 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.zon.completion;

import com.falsepattern.zigbrains.zon.parser.ZonFile;
import com.falsepattern.zigbrains.zon.psi.ZonProperty;
import com.falsepattern.zigbrains.zon.psi.ZonPropertyPlaceholder;
import com.falsepattern.zigbrains.zon.psi.ZonStruct;
import com.falsepattern.zigbrains.zon.psi.ZonTypes;
import com.falsepattern.zigbrains.zon.psi.impl.ZonPsiImplUtil;
import com.intellij.codeInsight.completion.CompletionContributor;
import com.intellij.codeInsight.completion.CompletionParameters;
import com.intellij.codeInsight.completion.CompletionProvider;
import com.intellij.codeInsight.completion.CompletionResultSet;
import com.intellij.codeInsight.completion.CompletionType;
import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.intellij.patterns.PlatformPatterns;
import com.intellij.util.ProcessingContext;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Set;

public class ZonCompletionContributor extends CompletionContributor {
    private static final List<String> ZON_ROOT_KEYS = List.of("name", "version", "dependencies");
    private static final List<String> ZON_DEP_KEYS = List.of("url", "hash");
    public ZonCompletionContributor() {
        extend(CompletionType.BASIC,
               PlatformPatterns.psiElement()
                               .withParent(PlatformPatterns.psiElement(ZonTypes.PROPERTY_PLACEHOLDER))
                               .withSuperParent(3, PlatformPatterns.psiElement(ZonFile.class)),
               new CompletionProvider<>() {
                    @Override
                    protected void addCompletions(@NotNull CompletionParameters parameters, @NotNull ProcessingContext context, @NotNull CompletionResultSet result) {
                        var placeholder = ZonPsiImplUtil.parent(parameters.getPosition(), ZonPropertyPlaceholder.class);
                        assert placeholder != null;

                        var zonStruct = ZonPsiImplUtil.parent(placeholder, ZonStruct.class);
                        assert zonStruct != null;
                        var keys = ZonPsiImplUtil.getKeys(zonStruct);
                        doAddCompletions(placeholder.getText().startsWith("."), keys, ZON_ROOT_KEYS, result);
                    }
                });
        extend(CompletionType.BASIC,
               PlatformPatterns.psiElement()
                               .withParent(PlatformPatterns.psiElement(ZonTypes.PROPERTY_PLACEHOLDER))
                               .withSuperParent(3, PlatformPatterns.psiElement(ZonTypes.PROPERTY))
                               .withSuperParent(5, PlatformPatterns.psiElement(ZonTypes.PROPERTY))
                               .withSuperParent(7, PlatformPatterns.psiElement(ZonFile.class)),
               new CompletionProvider<>() {
                   @Override
                   protected void addCompletions(@NotNull CompletionParameters parameters, @NotNull ProcessingContext context, @NotNull CompletionResultSet result) {
                       var placeholder = ZonPsiImplUtil.parent(parameters.getPosition(), ZonPropertyPlaceholder.class);
                       assert placeholder != null;
                       var depStruct = ZonPsiImplUtil.parent(placeholder, ZonStruct.class);
                       assert depStruct != null;
                       var parentProperty = ZonPsiImplUtil.parent(depStruct, ZonProperty.class);
                       assert parentProperty != null;
                       parentProperty = ZonPsiImplUtil.parent(parentProperty, ZonProperty.class);
                       assert parentProperty != null;
                       if (!"dependencies".equals(ZonPsiImplUtil.getText(parentProperty.getIdentifier()))) {
                           return;
                       }
                       doAddCompletions(placeholder.getText().startsWith("."), ZonPsiImplUtil.getKeys(depStruct), ZON_DEP_KEYS, result);
                   }
               });
    }
    private static void doAddCompletions(boolean hasDot, Set<String> current, List<String> expected, CompletionResultSet result) {
        for (var key: expected) {
            if (current.contains(key)) {
                continue;
            }
            result.addElement(LookupElementBuilder.create(hasDot ? key : "." + key));
        }
    }
}

File: core/src/main/kotlin/com/falsepattern/zigbrains/zon/completion/ZonCompletionContributor.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2025 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.zon.completion

import com.falsepattern.zigbrains.zon.psi.*
import com.intellij.codeInsight.completion.*
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.patterns.ElementPattern
import com.intellij.patterns.PlatformPatterns.psiElement
import com.intellij.patterns.PsiElementPattern
import com.intellij.psi.PsiElement
import com.intellij.psi.util.parentOfType
import com.intellij.util.ProcessingContext
import org.jetbrains.annotations.NonNls

class ZonCompletionContributor : CompletionContributor() {
    init {
        extend(
            CompletionType.BASIC,
            psiElement()
                .withParent(psiElement(ZonTypes.PROPERTY_PLACEHOLDER))
                .withSuperParent(4, psiOfType<ZonFile>())
        ) { parameters, _, result ->
            val placeholder = parameters.position.parentOfType<ZonPropertyPlaceholder>() ?: return@extend
            val zonEntry = placeholder.parentOfType<ZonEntry>() ?: return@extend
            val keys = zonEntry.keys
            doAddCompletions(placeholder.text.startsWith('.'), keys, ZON_ROOT_KEYS, result)
        }
        extend(
            CompletionType.BASIC,
            psiElement()
                .withParent(psiElement(ZonTypes.VALUE_PLACEHOLDER))
                .withSuperParent(2, psiElement(ZonTypes.LIST))
                .withSuperParent(4, psiOfType<ZonFile>())
        ) { _, _, result ->
            doAddCompletions(false, emptySet(), ZON_ROOT_KEYS, result)
        }
        extend(
            CompletionType.BASIC,
            psiElement()
                .withParent(psiElement(ZonTypes.PROPERTY_PLACEHOLDER))
                .withSuperParent(4, psiElement(ZonTypes.PROPERTY))
                .withSuperParent(7, psiElement(ZonTypes.PROPERTY))
                .withSuperParent(10, psiOfType<ZonFile>())
        ) { parameters, _, result ->
            val placeholder = parameters.position.parentOfType<ZonPropertyPlaceholder>() ?: return@extend
            val depEntry = placeholder.parentOfType<ZonEntry>() ?: return@extend
            if (depEntry.isDependency) {
                doAddCompletions(placeholder.text.startsWith('.'), emptySet(), ZON_DEP_KEYS, result)
            }
        }
        extend(
            CompletionType.BASIC,
            psiElement()
                .withParent(psiElement(ZonTypes.VALUE_PLACEHOLDER))
                .withSuperParent(2, psiElement(ZonTypes.LIST))
                .withSuperParent(4, psiElement(ZonTypes.PROPERTY))
                .withSuperParent(7, psiElement(ZonTypes.PROPERTY))
                .withSuperParent(10, psiOfType<ZonFile>())
        ) {parameters, _, result ->
            val placeholder = parameters.position.parentOfType<ZonValuePlaceholder>() ?: return@extend
            val depEntry = placeholder.parentOfType<ZonEntry>() ?: return@extend
            if (depEntry.isDependency) {
                doAddCompletions(false, emptySet(), ZON_DEP_KEYS, result)
            }
        }
        extend(
            CompletionType.BASIC,
            psiElement()
                .withParent(psiElement(ZonTypes.VALUE_PLACEHOLDER))
                .withSuperParent(5, psiElement(ZonTypes.PROPERTY))
                .withSuperParent(8, psiElement(ZonTypes.PROPERTY))
                .withSuperParent(11, psiOfType<ZonFile>())
        ) { parameters, _, result ->
            val placeholder = parameters.position.parentOfType<ZonValuePlaceholder>() ?: return@extend
            val valueProperty = placeholder.parentOfType<ZonProperty>() ?: return@extend
            if (valueProperty.isDependency && valueProperty.identifier.value == "lazy") {
                result.addElement(LookupElementBuilder.create("true"))
                result.addElement(LookupElementBuilder.create("false"))
            }
        }
    }

    private inline fun extend(type: CompletionType, place: ElementPattern<PsiElement>, crossinline action: Processor) {
        extend(type, place, object : CompletionProvider<CompletionParameters>() {
            override fun addCompletions(
                parameters: CompletionParameters, context: ProcessingContext, result: CompletionResultSet
            ) = action(parameters, context, result)
        })
    }
}

private typealias Processor = (
    parameters: CompletionParameters, context: ProcessingContext, result: CompletionResultSet
) -> Unit

@NonNls
private val ZON_ROOT_KEYS: List<String> =
    listOf("name", "version", "minimum_zig_version", "dependencies", "paths")
@NonNls
private val ZON_DEP_KEYS: List<String> = listOf("url", "hash", "path", "lazy")

private fun doAddCompletions(
    hasDot: Boolean, current: Set<String>, expected: List<String>, result: CompletionResultSet
) {
    for (key in expected) {
        if (current.contains(key)) {
            continue
        }
        result.addElement(LookupElementBuilder.create(if (hasDot) key else ".$key"))
    }
}

private val ZonProperty.isDependency: Boolean
    get() {
        return parentOfType<ZonEntry>()?.isDependency == true
    }

private val ZonEntry.isDependency: Boolean
    get() {
        val parentProperty = parentOfType<ZonProperty>()?.parentOfType<ZonProperty>() ?: return false
        return parentProperty.identifier.value == "dependencies"
    }

private inline fun <reified T : PsiElement> psiOfType(): PsiElementPattern.Capture<T> =
    psiElement(T::class.java)